<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.3.353">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Noor Sohail, Mary Piper, Lorena Pantano, Meeta Mistry, Radhika Khetani, Jihe Liu, Will Gammerdinger">

<title>Pseudobulk for single-cell RNA-seq - DE analysis using FindMarkers</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { display: inline-block; line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="../site_libs/quarto-nav/quarto-nav.js"></script>
<script src="../site_libs/quarto-nav/headroom.min.js"></script>
<script src="../site_libs/clipboard/clipboard.min.js"></script>
<script src="../site_libs/quarto-search/autocomplete.umd.js"></script>
<script src="../site_libs/quarto-search/fuse.min.js"></script>
<script src="../site_libs/quarto-search/quarto-search.js"></script>
<meta name="quarto:offset" content="../">
<script src="../site_libs/quarto-html/quarto.js"></script>
<script src="../site_libs/quarto-html/popper.min.js"></script>
<script src="../site_libs/quarto-html/tippy.umd.min.js"></script>
<script src="../site_libs/quarto-html/anchor.min.js"></script>
<link href="../site_libs/quarto-html/tippy.css" rel="stylesheet">
<link href="../site_libs/quarto-html/quarto-syntax-highlighting.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="../site_libs/bootstrap/bootstrap.min.js"></script>
<link href="../site_libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="../site_libs/bootstrap/bootstrap.min.css" rel="stylesheet" id="quarto-bootstrap" data-mode="light">
<script id="quarto-search-options" type="application/json">{
  "location": "navbar",
  "copy-button": false,
  "collapse-after": 3,
  "panel-placement": "end",
  "type": "overlay",
  "limit": 20,
  "language": {
    "search-no-results-text": "No results",
    "search-matching-documents-text": "matching documents",
    "search-copy-link-title": "Copy link to search",
    "search-hide-matches-text": "Hide additional matches",
    "search-more-match-text": "more match in this document",
    "search-more-matches-text": "more matches in this document",
    "search-clear-button-title": "Clear",
    "search-detached-cancel-button-title": "Cancel",
    "search-submit-button-title": "Submit"
  }
}</script>


<link rel="stylesheet" href="../css/styles.css">
</head>

<body class="nav-sidebar docked nav-fixed">

<div id="quarto-search-results"></div>
  <header id="quarto-header" class="headroom fixed-top">
    <nav class="navbar navbar-expand-lg navbar-dark ">
      <div class="navbar-container container-fluid">
      <div class="navbar-brand-container">
    <a class="navbar-brand" href="../index.html">
    <span class="navbar-title">Pseudobulk for single-cell RNA-seq</span>
    </a>
  </div>
            <div id="quarto-search" class="" title="Search"></div>
          <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbarCollapse" aria-controls="navbarCollapse" aria-expanded="false" aria-label="Toggle navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
  <span class="navbar-toggler-icon"></span>
</button>
          <div class="collapse navbar-collapse" id="navbarCollapse">
            <ul class="navbar-nav navbar-nav-scroll me-auto">
  <li class="nav-item">
    <a class="nav-link" href="../lessons/schedule.html" rel="" target="">
 <span class="menu-text">Schedule</span></a>
  </li>  
</ul>
            <ul class="navbar-nav navbar-nav-scroll ms-auto">
  <li class="nav-item">
    <a class="nav-link" href="https://bioinformatics.sph.harvard.edu/" rel="" target="">
 <span class="menu-text">HBC</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="https://github.com/hbctraining/Intro-to-scRNAseq-Quarto" rel="" target="">
 <span class="menu-text">GitHub</span></a>
  </li>  
  <li class="nav-item">
    <a class="nav-link" href="mailto:hbctraining@hsph.harvard.edu" rel="" target="">
 <span class="menu-text">Contact us</span></a>
  </li>  
</ul>
            <div class="quarto-navbar-tools">
</div>
          </div> <!-- /navcollapse -->
      </div> <!-- /container-fluid -->
    </nav>
  <nav class="quarto-secondary-nav">
    <div class="container-fluid d-flex">
      <button type="button" class="quarto-btn-toggle btn" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">
        <i class="bi bi-layout-text-sidebar-reverse"></i>
      </button>
      <nav class="quarto-page-breadcrumbs" aria-label="breadcrumb"><ol class="breadcrumb"><li class="breadcrumb-item"><a href="../lessons/01_setup_intro_dataset.html">Day 1:</a></li><li class="breadcrumb-item"><a href="../lessons/02_DEanalysis_using_FindMarkers.html">DE analysis using FindMarkers</a></li></ol></nav>
      <a class="flex-grow-1" role="button" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass" aria-controls="quarto-sidebar" aria-expanded="false" aria-label="Toggle sidebar navigation" onclick="if (window.quartoToggleHeadroom) { window.quartoToggleHeadroom(); }">      
      </a>
    </div>
  </nav>
</header>
<!-- content -->
<div id="quarto-content" class="quarto-container page-columns page-rows-contents page-layout-article page-navbar">
<!-- sidebar -->
  <nav id="quarto-sidebar" class="sidebar collapse collapse-horizontal sidebar-navigation docked overflow-auto">
    <div class="sidebar-menu-container"> 
    <ul class="list-unstyled mt-1">
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false">
 <span class="menu-text">Pre-reading:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-1" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-1" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/00_counts_to_clusters_overview.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Single-cell RNA-seq: From Counts to Clusters</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true">
 <span class="menu-text">Day 1:</span></a>
          <a class="sidebar-item-toggle text-start" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-2" aria-expanded="true" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-2" class="collapse list-unstyled sidebar-section depth1 show">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/01_setup_intro_dataset.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Project Setup and Data Exploration</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/02_DEanalysis_using_FindMarkers.html" class="sidebar-item-text sidebar-link active">
 <span class="menu-text">DE analysis using FindMarkers</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false">
 <span class="menu-text">Day 1 Self-learning:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-3" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-3" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/03_pseudobulk_DESeq2.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Set-up DESeq2 analysis</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/04_pseudobulk_DE_analysis.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pseudobulk DESeq2 analysis</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false">
 <span class="menu-text">Day 2:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-4" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-4" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/05_pseudobulk_DE_visualizations.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Pseudobulk visualization</span></a>
  </div>
</li>
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/06_DE_comparisons.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Comparing results from different DE approaches</span></a>
  </div>
</li>
      </ul>
  </li>
        <li class="sidebar-item sidebar-item-section">
      <div class="sidebar-item-container"> 
            <a class="sidebar-item-text sidebar-link text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false">
 <span class="menu-text">Day 2 Self-learning:</span></a>
          <a class="sidebar-item-toggle text-start collapsed" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar-section-5" aria-expanded="false" aria-label="Toggle section">
            <i class="bi bi-chevron-right ms-2"></i>
          </a> 
      </div>
      <ul id="quarto-sidebar-section-5" class="collapse list-unstyled sidebar-section depth1 ">  
          <li class="sidebar-item">
  <div class="sidebar-item-container"> 
  <a href="../lessons/07_functional_analysis_pseudobulk.html" class="sidebar-item-text sidebar-link">
 <span class="menu-text">Functional analysis of pseudobulk DE</span></a>
  </div>
</li>
      </ul>
  </li>
    </ul>
    </div>
</nav>
<div id="quarto-sidebar-glass" data-bs-toggle="collapse" data-bs-target="#quarto-sidebar,#quarto-sidebar-glass"></div>
<!-- margin-sidebar -->
    <div id="quarto-margin-sidebar" class="sidebar margin-sidebar">
        <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">On this page</h2>
   
  <ul>
  <li><a href="#learning-objectives" id="toc-learning-objectives" class="nav-link active" data-scroll-target="#learning-objectives">Learning Objectives:</a></li>
  <li><a href="#differential-expression-between-conditions-using-findmarkers" id="toc-differential-expression-between-conditions-using-findmarkers" class="nav-link" data-scroll-target="#differential-expression-between-conditions-using-findmarkers">Differential expression between conditions using FindMarkers()</a></li>
  <li><a href="#setting-up" id="toc-setting-up" class="nav-link" data-scroll-target="#setting-up">Setting up</a>
  <ul class="collapse">
  <li><a href="#load-libraries" id="toc-load-libraries" class="nav-link" data-scroll-target="#load-libraries">Load libraries</a></li>
  </ul></li>
  <li><a href="#findmarkers" id="toc-findmarkers" class="nav-link" data-scroll-target="#findmarkers">FindMarkers()</a></li>
  <li><a href="#significant-genes" id="toc-significant-genes" class="nav-link" data-scroll-target="#significant-genes">Significant genes</a>
  <ul class="collapse">
  <li><a href="#volcano-plot" id="toc-volcano-plot" class="nav-link" data-scroll-target="#volcano-plot">Volcano plot</a></li>
  <li><a href="#violin-plots" id="toc-violin-plots" class="nav-link" data-scroll-target="#violin-plots">Violin plots</a></li>
  <li><a href="#umap-plots" id="toc-umap-plots" class="nav-link" data-scroll-target="#umap-plots">UMAP plots</a></li>
  </ul></li>
  <li><a href="#other-statistical-tests" id="toc-other-statistical-tests" class="nav-link" data-scroll-target="#other-statistical-tests">Other statistical tests</a>
  <ul class="collapse">
  <li><a href="#wilcoxon-rank-sum-test" id="toc-wilcoxon-rank-sum-test" class="nav-link" data-scroll-target="#wilcoxon-rank-sum-test">Wilcoxon Rank Sum Test</a></li>
  <li><a href="#deseq2" id="toc-deseq2" class="nav-link" data-scroll-target="#deseq2">DESeq2</a></li>
  <li><a href="#mast" id="toc-mast" class="nav-link" data-scroll-target="#mast">MAST</a></li>
  </ul></li>
  </ul>
</nav>
    </div>
<!-- main -->
<main class="content" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<div class="quarto-title-block"><div><h1 class="title">DE analysis using FindMarkers</h1><button type="button" class="btn code-tools-button" id="quarto-code-tools-source"><i class="bi"></i> Code</button></div></div>
</div>



<div class="quarto-title-meta">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Noor Sohail, Mary Piper, Lorena Pantano, Meeta Mistry, Radhika Khetani, Jihe Liu, Will Gammerdinger </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">Invalid Date</p>
    </div>
  </div>
  
    
  </div>
  

</header>

<p>Approximate time: 75 minutes</p>
<section id="learning-objectives" class="level2">
<h2 class="anchored" data-anchor-id="learning-objectives">Learning Objectives:</h2>
<ul>
<li>Evaluate differential gene expression between conditions using a Wilcoxon rank sum test</li>
<li>Create visualizations for differentially expressed genes</li>
<li>Discuss other statistical tests for differential expression analysis</li>
</ul>
</section>
<section id="differential-expression-between-conditions-using-findmarkers" class="level2">
<h2 class="anchored" data-anchor-id="differential-expression-between-conditions-using-findmarkers">Differential expression between conditions using FindMarkers()</h2>
<p>In our current UMAP, we have merged samples across the different conditions and used integration to align cells of the same celltype across samples. Now, what if we were interested in a particular celltype and <strong>understanding how gene expression changes across the different conditions?</strong></p>
<p>The <code>FindMarkers()</code> function in the Seurat package is used to perform differential expression analysis between groups of cells. We provide arguments to specify <code>ident.1</code> and <code>ident.2</code> the two groups of cells we are interested in comparing. This function is commonly used at the <a href="https://hbctraining.github.io/Intro-to-scRNAseq/lessons/09_merged_SC_marker_identification.html">celltype annotation step</a>, but can also be used to calculate differentially expressed genes between experimental conditions.</p>
<p align="center">
<img src="../img/marker_ident_function3.png" width="350">
</p>
<p>We can use this same function to compare two groups of cells which represent different conditions by modifying the <code>ident</code> values provided. The <code>FindMarkers()</code> function has <strong>several important arguments</strong> we can modify when running it. To view what these parameters are, we can access the help page on this function:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>?FindMarkers</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Below we have described in more detail what each of these arguments mean :</p>
<ul>
<li><code>logfc.threshold</code>: The minimum log2 fold change for average expression of gene in group relative to the average expression in all other groups combined. Default is 0.25.</li>
<li><code>min.diff.pct</code>: The minimum percent difference between the percent of cells expressing the gene in the group and the percent of cells expressing gene in all other groups combined.</li>
<li><code>min.pct</code>: Only test genes that are detected in a minimum fraction of cells in either of the two populations. Meant to speed up the function by not testing genes that are very infrequently expressed. Default is 0.1.</li>
<li><code>ident.1</code>: This function only evaluates one group at a time; here you would specify the group of interest. These values must be set in the <code>Idents()</code>.</li>
<li><code>ident.2</code>: Here you would specify the group you want to compare <code>ident.1</code> against.</li>
</ul>
</section>
<section id="setting-up" class="level2">
<h2 class="anchored" data-anchor-id="setting-up">Setting up</h2>
<p>Open up a new Rscript file, and start with some comments to indicate what this file is going to contain:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Single-cell RNA-seq DE analysis - FindMarkers</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Save the Rscript as findMarkers.R.</p>
<section id="load-libraries" class="level3">
<h3 class="anchored" data-anchor-id="load-libraries">Load libraries</h3>
<p>Now let’s load the required libraries for this analysis:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Load libraries</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(EnhancedVolcano)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
</section>
<section id="findmarkers" class="level2">
<h2 class="anchored" data-anchor-id="findmarkers">FindMarkers()</h2>
<p>To use the function to look for DE genes between conditions, there are two things we need to do:</p>
<ol type="1">
<li>Subset the Seurat object to our celltype of interest</li>
<li>Set our active Idents to be the metadata column which specifies what condition each cell is</li>
</ol>
<p>In our example we are focusing on vascular smooth muscle (<code>VSM</code>) cells. The comparison we will be making is <code>TN</code> vs.&nbsp;<code>cold7</code>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset the object</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a>seurat_vsm <span class="ot">&lt;-</span> <span class="fu">subset</span>(seurat, <span class="at">subset =</span> (celltype <span class="sc">==</span> <span class="st">"VSM"</span>))</span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a><span class="co"># Set the idents</span></span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(seurat_vsm) <span class="ot">&lt;-</span> <span class="st">"condition"</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>seurat_vsm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>An object of class Seurat 
40338 features across 12372 samples within 3 assays 
Active assay: RNA (19771 features, 3000 variable features)
 2 layers present: data, counts
 2 other assays present: SCT, integrated
 2 dimensional reductions calculated: pca, umap</code></pre>
</div>
</div>
<p>Before we make our comparisons we will explicitly set our default assay, we want to use the <strong>normalized data, not the integrated data</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb6"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set default assay</span></span>
<span id="cb6-2"><a href="#cb6-2" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(seurat_vsm) <span class="ot">&lt;-</span> <span class="st">"RNA"</span></span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>The default assay should have already been <code>RNA</code>, but we encourage you to run this line of code above to be absolutely sure in case the active slot was changed somewhere upstream in your analysis.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Why don’t we use SCT normalized data?
</div>
</div>
<div class="callout-body-container callout-body">
<p>Note that the raw and normalized counts are stored in the <code>counts</code> and <code>data</code> slots of <code>RNA</code> assay, respectively. By default, the functions for finding markers will use normalized data if RNA is the DefaultAssay. The number of features in the <code>RNA</code> assay corresponds to all genes in our dataset.</p>
<p>Now if we consider the <code>SCT</code> assay, functions for finding markers would use the <code>scale.data</code> slot which is the pearson residuals that come out of regularized NB regression. Differential expression on these values can be difficult interpret. Additionally, only the variable features are represented in this assay and so we may not have data for some of our marker genes.</p>
</div>
</div>
<p>Now we can run <code>FindMarkers()</code>:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb7"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine differentiating markers for TN and cold7</span></span>
<span id="cb7-2"><a href="#cb7-2" aria-hidden="true" tabindex="-1"></a>dge_vsm <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(seurat_vsm,</span>
<span id="cb7-3"><a href="#cb7-3" aria-hidden="true" tabindex="-1"></a>                       <span class="at">ident.1=</span><span class="st">"cold7"</span>,</span>
<span id="cb7-4"><a href="#cb7-4" aria-hidden="true" tabindex="-1"></a>                       <span class="at">ident.2=</span><span class="st">"TN"</span></span>
<span id="cb7-5"><a href="#cb7-5" aria-hidden="true" tabindex="-1"></a>                       )</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
<p>Now let’s take a quick look at the results:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb8"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" aria-hidden="true" tabindex="-1"></a><span class="co"># View results </span></span>
<span id="cb8-2"><a href="#cb8-2" aria-hidden="true" tabindex="-1"></a>dge_vsm <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                p_val avg_log2FC pct.1 pct.2     p_val_adj
Gm42418 4.492261e-223  1.8013603 0.999 0.986 8.881650e-219
Ubb     1.801612e-207 -0.8274973 0.982 0.999 3.561968e-203
H3f3b   6.044423e-198 -0.9395566 0.975 0.998 1.195043e-193
Nr4a2   9.607598e-193 -2.1841673 0.503 0.855 1.899518e-188
Rpl21   2.319987e-190 -0.6377144 0.967 0.998 4.586847e-186
Rpl9    1.249444e-187 -0.6824496 0.954 0.999 2.470276e-183</code></pre>
</div>
</div>
<p><strong>The output from the <code>FindMarkers()</code> function</strong> is a matrix containing a ranked list of differentially expressed genes listed by gene ID and associated statistics. We describe some of these columns below:</p>
<ul>
<li><strong>gene:</strong> The gene symbol</li>
<li><strong>p_val:</strong> The p-value not adjusted for multiple test corrections</li>
<li><strong>avg_logFC:</strong> The average log-fold change between sample groups. Positive values indicate that the gene is more highly expressed <code>ident.1</code>.<br>
</li>
<li><strong>pct.1:</strong> The percentage of cells where the gene is detected in <code>ident.1</code> (cold7)<br>
</li>
<li><strong>pct.2:</strong> The percentage of cells where the gene is detected on average in <code>ident.2</code> (TN)</li>
<li><strong>p_val_adj:</strong> The adjusted p-value, based on Bonferroni correction using all genes in the dataset, used to determine significance</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>These tests treat each cell as an independent replicate and ignore inherent correlations between cells originating from the same sample. This results in highly <strong>inflated p-values</strong> for each gene. Studies have been shown to find a large number of false positive associations with these results.</p>
</div>
</div>
<p>When looking at the output, <strong>we suggest looking for markers with large differences in expression between <code>pct.1</code> and <code>pct.2</code> and larger fold changes</strong>. For instance if <code>pct.1</code> = 0.90 and <code>pct.2</code> = 0.80, it may not be as exciting of a marker. However, if <code>pct.2</code> = 0.1 instead, the bigger difference would be more convincing. Also of interest is whether the majority of cells expressing the gene are in the group of interest. If <code>pct.1</code> is low, such as 0.3, it may not be as interesting. Both of these are also possible parameters to include when running the function, as described above.</p>
<p>This is a great spot to pause and save your results!</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb10"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb10-1"><a href="#cb10-1" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(dge_vsm, <span class="st">"../results/findmarkers_vsm_cold7_vs_TN.csv"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div>
</section>
<section id="significant-genes" class="level2">
<h2 class="anchored" data-anchor-id="significant-genes">Significant genes</h2>
<p>We want to subset our results to show just our significant genes so we can begin visualizing and analyzing the results. To do this, we filter out rows based upon the <code>p_val_adj</code> column and subsetting only genes that meet our multiple testing-corrected significance threshold of 0.05.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb11"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb11-1"><a href="#cb11-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset significant genes</span></span>
<span id="cb11-2"><a href="#cb11-2" aria-hidden="true" tabindex="-1"></a>dge_vsm_sig <span class="ot">&lt;-</span> dge_vsm <span class="sc">%&gt;%</span> <span class="fu">subset</span>(p_val_adj <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span>
<span id="cb11-3"><a href="#cb11-3" aria-hidden="true" tabindex="-1"></a>dge_vsm_sig <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>                p_val avg_log2FC pct.1 pct.2     p_val_adj
Gm42418 4.492261e-223  1.8013603 0.999 0.986 8.881650e-219
Ubb     1.801612e-207 -0.8274973 0.982 0.999 3.561968e-203
H3f3b   6.044423e-198 -0.9395566 0.975 0.998 1.195043e-193
Nr4a2   9.607598e-193 -2.1841673 0.503 0.855 1.899518e-188
Rpl21   2.319987e-190 -0.6377144 0.967 0.998 4.586847e-186
Rpl9    1.249444e-187 -0.6824496 0.954 0.999 2.470276e-183</code></pre>
</div>
</div>
<section id="volcano-plot" class="level3">
<h3 class="anchored" data-anchor-id="volcano-plot">Volcano plot</h3>
<p>To get a first look at the genes that are retained, we can generated a volcano plot using the <code>EnhancedVolcano()</code> function. This is a visualization that allows us to quickly see trends in the significant genes. The x-axis here represents the average log2-fold change value, showing the degree of difference between the two conditions. On the y-axis, we see our <code>p_val_adj</code> column represented after a negative log10 transformation is applied to better see the spread of the adjusted p-values.</p>
<p>Volcano plots provide a great <strong>overview of which genes are up-regulated (positive on the x-axis) or down-regulated (negative on the x-axis)</strong>.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb13"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb13-1"><a href="#cb13-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Volcano plot</span></span>
<span id="cb13-2"><a href="#cb13-2" aria-hidden="true" tabindex="-1"></a>p_fm <span class="ot">&lt;-</span> <span class="fu">EnhancedVolcano</span>(dge_vsm_sig,</span>
<span id="cb13-3"><a href="#cb13-3" aria-hidden="true" tabindex="-1"></a>        <span class="fu">row.names</span>(dge_vsm_sig),</span>
<span id="cb13-4"><a href="#cb13-4" aria-hidden="true" tabindex="-1"></a>        <span class="at">x=</span><span class="st">"avg_log2FC"</span>,</span>
<span id="cb13-5"><a href="#cb13-5" aria-hidden="true" tabindex="-1"></a>        <span class="at">y=</span><span class="st">"p_val_adj"</span>,</span>
<span id="cb13-6"><a href="#cb13-6" aria-hidden="true" tabindex="-1"></a>       <span class="at">title=</span><span class="st">"FindMarkers VSM cells"</span>,</span>
<span id="cb13-7"><a href="#cb13-7" aria-hidden="true" tabindex="-1"></a>       <span class="at">subtitle=</span><span class="st">"TN vs cold7"</span>)</span>
<span id="cb13-8"><a href="#cb13-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb13-9"><a href="#cb13-9" aria-hidden="true" tabindex="-1"></a>p_fm</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_DEanalysis_using_FindMarkers_files/figure-html/unnamed-chunk-11-1.png" class="img-fluid" width="960"></p>
</div>
</div>
</section>
<section id="violin-plots" class="level3">
<h3 class="anchored" data-anchor-id="violin-plots">Violin plots</h3>
<p>While looking at the overall trends in the data is a great starting point, we can also start looking at genes that have large differences between <code>TN</code> and <code>cold7</code>. To do this, we can take a look at the top 6 genes with the smallest p-values. We additionally disregard the ribsomal genes in this visualization step.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb14"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb14-1"><a href="#cb14-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the gene names and get the first 6 values</span></span>
<span id="cb14-2"><a href="#cb14-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Ignore ribosomal genes</span></span>
<span id="cb14-3"><a href="#cb14-3" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> dge_vsm_sig <span class="sc">%&gt;%</span></span>
<span id="cb14-4"><a href="#cb14-4" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var=</span><span class="st">"gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb14-5"><a href="#cb14-5" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">str_detect</span>(gene, <span class="st">"Rpl|Rps"</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb14-6"><a href="#cb14-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>(<span class="dv">6</span>)</span>
<span id="cb14-7"><a href="#cb14-7" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> genes<span class="sc">$</span>gene</span>
<span id="cb14-8"><a href="#cb14-8" aria-hidden="true" tabindex="-1"></a>genes</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output cell-output-stdout">
<pre><code>[1] "Gm42418" "Ubb"     "H3f3b"   "Nr4a2"   "Cebpb"   "Fau"    </code></pre>
</div>
</div>
<p>With these genes selected, we can now being to visualize the distribution of expression across our two conditions using the <code>VlnPlot()</code> function.</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb16"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb16-1"><a href="#cb16-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Set Idents and draw Violin plots for top 6 genes</span></span>
<span id="cb16-2"><a href="#cb16-2" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(seurat_vsm) <span class="ot">&lt;-</span> <span class="st">"condition"</span></span>
<span id="cb16-3"><a href="#cb16-3" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(seurat_vsm, genes, <span class="at">ncol=</span><span class="dv">3</span>, <span class="at">idents=</span><span class="fu">c</span>(<span class="st">"TN"</span>, <span class="st">"cold7"</span>))</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_DEanalysis_using_FindMarkers_files/figure-html/unnamed-chunk-13-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
<section id="umap-plots" class="level3">
<h3 class="anchored" data-anchor-id="umap-plots">UMAP plots</h3>
<p>When comparing two different conditions, we recommend creating a UMAP that clearly shows where the cells exist for each condition. To do so, we first need to get the UMAP coordinates for every cell of interest. When creating the scatterplot, the first thing we do is put a layer of light gray points that show the entire dataset to understand where all the cells fall. Then, we take the UMAP coordinates of the condition (<code>TN</code> or <code>cold7</code> in our example) and plot those on top with a color to clearly indicate where those cells are located.</p>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>This sometimes works better on the non-integrated data, so you observe a true separation of cells by condition.</p>
</div>
</div>
<div class="cell">
<div class="sourceCode cell-code" id="cb17"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb17-1"><a href="#cb17-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Grab the umap coordinates and condition information for each cell</span></span>
<span id="cb17-2"><a href="#cb17-2" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(seurat_vsm, <span class="fu">c</span>(<span class="st">"umap_1"</span>, <span class="st">"umap_2"</span>, <span class="st">"condition"</span>))</span>
<span id="cb17-3"><a href="#cb17-3" aria-hidden="true" tabindex="-1"></a>df_tn <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span> <span class="fu">subset</span>(condition <span class="sc">==</span> <span class="st">"TN"</span>)</span>
<span id="cb17-4"><a href="#cb17-4" aria-hidden="true" tabindex="-1"></a>df_cold7 <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span> <span class="fu">subset</span>(condition <span class="sc">==</span> <span class="st">"cold7"</span>)</span>
<span id="cb17-5"><a href="#cb17-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-6"><a href="#cb17-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Scatterplot of TN cells</span></span>
<span id="cb17-7"><a href="#cb17-7" aria-hidden="true" tabindex="-1"></a>p_tn <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb17-8"><a href="#cb17-8" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>df, <span class="fu">aes</span>(<span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2), <span class="at">color=</span><span class="st">"lightgray"</span>, <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">size =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb17-9"><a href="#cb17-9" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>df_tn, <span class="fu">aes</span>(<span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2), <span class="at">color=</span><span class="st">"#F8766D"</span>, <span class="at">size =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb17-10"><a href="#cb17-10" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb17-11"><a href="#cb17-11" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"VSM: TN cells"</span>)</span>
<span id="cb17-12"><a href="#cb17-12" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-13"><a href="#cb17-13" aria-hidden="true" tabindex="-1"></a><span class="co"># Scatterplot of cold7 cells</span></span>
<span id="cb17-14"><a href="#cb17-14" aria-hidden="true" tabindex="-1"></a>p_cold7 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb17-15"><a href="#cb17-15" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>df, <span class="fu">aes</span>(<span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2), <span class="at">color=</span><span class="st">"lightgray"</span>, <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">size =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb17-16"><a href="#cb17-16" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>df_cold7, <span class="fu">aes</span>(<span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2), <span class="at">color=</span><span class="st">"#00B8E7"</span>, <span class="at">size =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb17-17"><a href="#cb17-17" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb17-18"><a href="#cb17-18" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"VSM: cold7 cells"</span>)</span>
<span id="cb17-19"><a href="#cb17-19" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb17-20"><a href="#cb17-20" aria-hidden="true" tabindex="-1"></a><span class="co"># TN and cold7 UMAPs side by side</span></span>
<span id="cb17-21"><a href="#cb17-21" aria-hidden="true" tabindex="-1"></a>p_tn <span class="sc">+</span> p_cold7</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_DEanalysis_using_FindMarkers_files/figure-html/unnamed-chunk-14-1.png" class="img-fluid" width="672"></p>
</div>
</div>
<p>This allows us to better understand our results when we look at any follow-up information on our UMAP. For example, we can begin to look at the distribution of gene expression for each of the top 6 genes with a better understanding of where the cells for each condition lie:</p>
<div class="cell">
<div class="sourceCode cell-code" id="cb18"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb18-1"><a href="#cb18-1" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seurat_vsm, genes, <span class="at">ncol=</span><span class="dv">3</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
<div class="cell-output-display">
<p><img src="02_DEanalysis_using_FindMarkers_files/figure-html/unnamed-chunk-15-1.png" class="img-fluid" width="672"></p>
</div>
</div>
</section>
</section>
<section id="other-statistical-tests" class="level2">
<h2 class="anchored" data-anchor-id="other-statistical-tests">Other statistical tests</h2>
<p>When we looked at the extra explanations for the <code>FindMarkers()</code> function, there was a parameter called <code>test.use</code>. By default, the method for calculating differentially expressed genes will be a Wilcoxon Rank Sum Test. This is a fairly simple statistical approach, and there a multitude of different algorithms that can be specified. These other options are documented on the <code>FindMarkers()</code> <a href="https://www.rdocumentation.org/packages/Seurat/versions/5.0.3/topics/FindMarkers">documentation page</a>. For this workshop we want to highlight a few of these methods:</p>
<section id="wilcoxon-rank-sum-test" class="level3">
<h3 class="anchored" data-anchor-id="wilcoxon-rank-sum-test">Wilcoxon Rank Sum Test</h3>
<ul>
<li>Often described as the non-parametric version of the two-sample t-test.</li>
<li>Beneficial because it can reduce the impact of outliers, which can skew the results of parametric testing.</li>
<li>It ranks the data and compares the sum of ranks within each group, to identify significant differences.</li>
</ul>
</section>
<section id="deseq2" class="level3">
<h3 class="anchored" data-anchor-id="deseq2">DESeq2</h3>
<ul>
<li>Identifies differentially expressed genes between two groups of cells based on a model using DESeq2 which uses a negative binomial distribution (<a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8">Love et al., 2014</a>). More information on DESeq2 will be provided in an upcoming lesson in this workshop.</li>
<li>This test option does not support pre-filtering of genes based on average difference (or percent detection rate) between cell groups. However, genes may be pre-filtered based on their minimum detection rate (min.pct) across both cell groups.</li>
</ul>
<div class="callout callout-style-default callout-note callout-titled">
<div class="callout-header d-flex align-content-center">
<div class="callout-icon-container">
<i class="callout-icon"></i>
</div>
<div class="callout-title-container flex-fill">
Note
</div>
</div>
<div class="callout-body-container callout-body">
<p>The creators of the Seurat package <a href="https://github.com/satijalab/seurat/issues/2938">no longer recommend</a> using the <code>FindMarkers()</code> implementation of DESeq2.</p>
</div>
</div>
</section>
<section id="mast" class="level3">
<h3 class="anchored" data-anchor-id="mast">MAST</h3>
<ul>
<li>Implements an approach that accounts for the stochastic dropout and characteristic bimodal expression distributions in which expression is either strongly non-zero or non-detectable.
<ul>
<li>A two-part, generalized linear model for such bimodal data that parameterizes both of these features<br>
</li>
</ul></li>
<li>Also allows for estimation and control of the “cellular detection rate” (CDR) while simultaneously estimating treatment effects. This addresses the fact that cells scale transcript copy number with cell volume.</li>
<li>Permits the analysis of complex experiments, such as repeated single-cell measurements under various treatments or longitudinal sampling of single cells from multiple subjects with a variety of background characteristics (e.g., sex, age), because it can easily be extended to accommodate random effects.</li>
</ul>
<blockquote class="blockquote">
<p><strong>NOTE:</strong> Instead of using the <code>FindMarkers()</code> implementation, we recommend directly using the <a href="https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0844-5">MAST</a> algorithm from the package itself for the best results.</p>
If you are interested in exploring code to run MAST on this dataset directly using the package, please see the script at the link below. We recommend including the sample in the model to improve results by taking into account biological variability. Please note that this is a <strong>computationally intensive</strong> calculation and may take <strong>a long time to run</strong>.
<details>
<summary>
<b><i>Click here for code to run MAST on this dataset side-by-side</i></b>
</summary>
<br>Note that this R code below uses the <b>MAST library</b>. In order to run this you will need to first install the required packagea and then &gt; load the libraries.

<pre>library(Seurat)
library(dplyr)
library(SingleCellExperiment)
library(MAST)

# Seurat to SingleCellExperiment
DefaultAssay(seurat_vsm) &lt;- "RNA"
sce &lt;- as.SingleCellExperiment(seurat_vsm)
# Apply log transformation
assay(sce, "logcounts") &lt;- log2(counts(sce) + 1)

# Create new sce object (with only 'logcounts' matrix)
sce_1 &lt;- SingleCellExperiment(assays = list(logcounts = assay(sce, "logcounts")))
colData(sce_1) &lt;- colData(sce)

# Change to SingleCellAssay
sca &lt;- SceToSingleCellAssay(sce_1)

# Calculate number of genes expressed per cell and scale the value
cdr2 &lt;- colSums(SummarizedExperiment::assay(sca) &gt; 0)
colData(sca)$cngeneson &lt;- scale(cdr2)

# Takes a long time to calculate!
# Here our model includes:
	# The number of genes expressed (ngeneson)
	# Experimental condition (condition)
	# Sample as a random variable ((1 | sample))

zlmCond &lt;- zlm(~condition + cngeneson + (1 | sample), 
               sca, method="glmer", ebayes=FALSE)

# Only test the condition coefficient.
summaryCond &lt;- summary(zlmCond, doLRT='conditionTN') 

# Some data wranging of the results
summaryDt &lt;- summaryCond$datatable
fcHurdle &lt;- merge(summaryDt[contrast=='conditionTN' &amp; component=='H',.(primerid, `Pr(&gt;Chisq)`)], #hurdle P values
                 summaryDt[contrast=='conditionTN' &amp; component=='logFC', 
                           .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients

fcHurdle[,fdr:=p.adjust(`Pr(&gt;Chisq)`, 'fdr')]
</pre>
</details>
</blockquote>


<!-- -->

</section>
</section>

</main> <!-- /main -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button', {
    text: function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
    }
  });
  clipboard.on('success', function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  });
  const viewSource = window.document.getElementById('quarto-view-source') ||
                     window.document.getElementById('quarto-code-tools-source');
  if (viewSource) {
    const sourceUrl = viewSource.getAttribute("data-quarto-source-url");
    viewSource.addEventListener("click", function(e) {
      if (sourceUrl) {
        // rstudio viewer pane
        if (/\bcapabilities=\b/.test(window.location)) {
          window.open(sourceUrl);
        } else {
          window.location.href = sourceUrl;
        }
      } else {
        const modal = new bootstrap.Modal(document.getElementById('quarto-embedded-source-code-modal'));
        modal.show();
      }
      return false;
    });
  }
  function toggleCodeHandler(show) {
    return function(e) {
      const detailsSrc = window.document.querySelectorAll(".cell > details > .sourceCode");
      for (let i=0; i<detailsSrc.length; i++) {
        const details = detailsSrc[i].parentElement;
        if (show) {
          details.open = true;
        } else {
          details.removeAttribute("open");
        }
      }
      const cellCodeDivs = window.document.querySelectorAll(".cell > .sourceCode");
      const fromCls = show ? "hidden" : "unhidden";
      const toCls = show ? "unhidden" : "hidden";
      for (let i=0; i<cellCodeDivs.length; i++) {
        const codeDiv = cellCodeDivs[i];
        if (codeDiv.classList.contains(fromCls)) {
          codeDiv.classList.remove(fromCls);
          codeDiv.classList.add(toCls);
        } 
      }
      return false;
    }
  }
  const hideAllCode = window.document.getElementById("quarto-hide-all-code");
  if (hideAllCode) {
    hideAllCode.addEventListener("click", toggleCodeHandler(false));
  }
  const showAllCode = window.document.getElementById("quarto-show-all-code");
  if (showAllCode) {
    showAllCode.addEventListener("click", toggleCodeHandler(true));
  }
  function tippyHover(el, contentFn) {
    const config = {
      allowHTML: true,
      content: contentFn,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start'
    };
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      return note.innerHTML;
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script><div class="modal fade" id="quarto-embedded-source-code-modal" tabindex="-1" aria-labelledby="quarto-embedded-source-code-modal-label" aria-hidden="true"><div class="modal-dialog modal-dialog-scrollable"><div class="modal-content"><div class="modal-header"><h5 class="modal-title" id="quarto-embedded-source-code-modal-label">Source Code</h5><button class="btn-close" data-bs-dismiss="modal"></button></div><div class="modal-body"><div class="">
<div class="sourceCode" id="cb19" data-shortcodes="false"><pre class="sourceCode markdown code-with-copy"><code class="sourceCode markdown"><span id="cb19-1"><a href="#cb19-1" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb19-2"><a href="#cb19-2" aria-hidden="true" tabindex="-1"></a><span class="an">title:</span><span class="co"> "DE analysis using FindMarkers"</span></span>
<span id="cb19-3"><a href="#cb19-3" aria-hidden="true" tabindex="-1"></a><span class="an">author:</span><span class="co"> "Noor Sohail, Mary Piper, Lorena Pantano, Meeta Mistry, Radhika Khetani, Jihe Liu, Will Gammerdinger"</span></span>
<span id="cb19-4"><a href="#cb19-4" aria-hidden="true" tabindex="-1"></a><span class="an">date:</span><span class="co"> Wednesday, February 5th, 2020</span></span>
<span id="cb19-5"><a href="#cb19-5" aria-hidden="true" tabindex="-1"></a><span class="co">---</span></span>
<span id="cb19-6"><a href="#cb19-6" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-7"><a href="#cb19-7" aria-hidden="true" tabindex="-1"></a>Approximate time: 75 minutes</span>
<span id="cb19-8"><a href="#cb19-8" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-11"><a href="#cb19-11" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb19-12"><a href="#cb19-12" aria-hidden="true" tabindex="-1"></a><span class="co">#| echo: false</span></span>
<span id="cb19-13"><a href="#cb19-13" aria-hidden="true" tabindex="-1"></a>seurat <span class="ot">&lt;-</span> <span class="fu">readRDS</span>(<span class="st">"../data/BAT_GSE160585_final.rds"</span>)</span>
<span id="cb19-14"><a href="#cb19-14" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb19-15"><a href="#cb19-15" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-16"><a href="#cb19-16" aria-hidden="true" tabindex="-1"></a><span class="fu">## Learning Objectives:</span></span>
<span id="cb19-17"><a href="#cb19-17" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-18"><a href="#cb19-18" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Evaluate differential gene expression between conditions using a Wilcoxon rank sum test</span>
<span id="cb19-19"><a href="#cb19-19" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Create visualizations for differentially expressed genes</span>
<span id="cb19-20"><a href="#cb19-20" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Discuss other statistical tests for differential expression analysis</span>
<span id="cb19-21"><a href="#cb19-21" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-22"><a href="#cb19-22" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-23"><a href="#cb19-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-24"><a href="#cb19-24" aria-hidden="true" tabindex="-1"></a><span class="fu">## Differential expression between conditions using FindMarkers()</span></span>
<span id="cb19-25"><a href="#cb19-25" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-26"><a href="#cb19-26" aria-hidden="true" tabindex="-1"></a>In our current UMAP, we have merged samples across the different conditions and used integration to align cells of the same celltype across samples. Now, what if we were interested in a particular celltype and **understanding how gene expression changes across the different conditions?**</span>
<span id="cb19-27"><a href="#cb19-27" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-28"><a href="#cb19-28" aria-hidden="true" tabindex="-1"></a>The <span class="in">`FindMarkers()`</span> function in the Seurat package is used to perform differential expression analysis between groups of cells. We provide arguments to specify <span class="in">`ident.1`</span> and <span class="in">`ident.2`</span> the two groups of cells we are interested in comparing. This function is commonly used at the <span class="co">[</span><span class="ot">celltype annotation step</span><span class="co">](https://hbctraining.github.io/Intro-to-scRNAseq/lessons/09_merged_SC_marker_identification.html)</span>, but can also be used to calculate differentially expressed genes between experimental conditions.</span>
<span id="cb19-29"><a href="#cb19-29" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-30"><a href="#cb19-30" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;p</span> <span class="er">align</span><span class="ot">=</span><span class="st">"center"</span><span class="kw">&gt;</span></span>
<span id="cb19-31"><a href="#cb19-31" aria-hidden="true" tabindex="-1"></a>    <span class="kw">&lt;img</span> <span class="er">src</span><span class="ot">=</span><span class="st">"../img/marker_ident_function3.png"</span> <span class="er">width</span><span class="ot">=</span><span class="st">"350"</span><span class="kw">&gt;</span></span>
<span id="cb19-32"><a href="#cb19-32" aria-hidden="true" tabindex="-1"></a><span class="kw">&lt;/p&gt;</span></span>
<span id="cb19-33"><a href="#cb19-33" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-34"><a href="#cb19-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-35"><a href="#cb19-35" aria-hidden="true" tabindex="-1"></a>We can use this same function to compare two groups of cells which represent different conditions by modifying the <span class="in">`ident`</span> values provided. The <span class="in">`FindMarkers()`</span> function has **several important arguments** we can modify when running it. To view what these parameters are, we can access the help page on this function:</span>
<span id="cb19-36"><a href="#cb19-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-39"><a href="#cb19-39" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb19-40"><a href="#cb19-40" aria-hidden="true" tabindex="-1"></a><span class="co">#| eval: false</span></span>
<span id="cb19-41"><a href="#cb19-41" aria-hidden="true" tabindex="-1"></a>?FindMarkers</span>
<span id="cb19-42"><a href="#cb19-42" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb19-43"><a href="#cb19-43" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-44"><a href="#cb19-44" aria-hidden="true" tabindex="-1"></a>Below we have described in more detail what each of these arguments mean :</span>
<span id="cb19-45"><a href="#cb19-45" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-46"><a href="#cb19-46" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`logfc.threshold`</span>: The minimum log2 fold change for average expression of gene in group relative to the average expression in all other groups combined. Default is 0.25.</span>
<span id="cb19-47"><a href="#cb19-47" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`min.diff.pct`</span>: The minimum percent difference between the percent of cells expressing the gene in the group and the percent of cells expressing gene in all other groups combined.</span>
<span id="cb19-48"><a href="#cb19-48" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`min.pct`</span>: Only test genes that are detected in a minimum fraction of cells in either of the two populations. Meant to speed up the function by not testing genes that are very infrequently expressed. Default is 0.1.</span>
<span id="cb19-49"><a href="#cb19-49" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`ident.1`</span>: This function only evaluates one group at a time; here you would specify the group of interest. These values must be set in the <span class="in">`Idents()`</span>.</span>
<span id="cb19-50"><a href="#cb19-50" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span><span class="in">`ident.2`</span>: Here you would specify the group you want to compare <span class="in">`ident.1`</span> against.</span>
<span id="cb19-51"><a href="#cb19-51" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-52"><a href="#cb19-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-53"><a href="#cb19-53" aria-hidden="true" tabindex="-1"></a><span class="fu">## Setting up </span></span>
<span id="cb19-54"><a href="#cb19-54" aria-hidden="true" tabindex="-1"></a>Open up a new Rscript file, and start with some comments to indicate what this file is going to contain:</span>
<span id="cb19-55"><a href="#cb19-55" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-58"><a href="#cb19-58" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb19-59"><a href="#cb19-59" aria-hidden="true" tabindex="-1"></a><span class="co"># Single-cell RNA-seq DE analysis - FindMarkers</span></span>
<span id="cb19-60"><a href="#cb19-60" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb19-61"><a href="#cb19-61" aria-hidden="true" tabindex="-1"></a>Save the Rscript as findMarkers.R. </span>
<span id="cb19-62"><a href="#cb19-62" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-63"><a href="#cb19-63" aria-hidden="true" tabindex="-1"></a><span class="fu">### Load libraries</span></span>
<span id="cb19-64"><a href="#cb19-64" aria-hidden="true" tabindex="-1"></a>Now let's load the required libraries for this analysis:</span>
<span id="cb19-65"><a href="#cb19-65" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-68"><a href="#cb19-68" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb19-69"><a href="#cb19-69" aria-hidden="true" tabindex="-1"></a><span class="co"># Load libraries</span></span>
<span id="cb19-70"><a href="#cb19-70" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(Seurat)</span>
<span id="cb19-71"><a href="#cb19-71" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb19-72"><a href="#cb19-72" aria-hidden="true" tabindex="-1"></a><span class="fu">library</span>(EnhancedVolcano)</span>
<span id="cb19-73"><a href="#cb19-73" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb19-74"><a href="#cb19-74" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-75"><a href="#cb19-75" aria-hidden="true" tabindex="-1"></a><span class="fu">## FindMarkers()</span></span>
<span id="cb19-76"><a href="#cb19-76" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-77"><a href="#cb19-77" aria-hidden="true" tabindex="-1"></a>To use the function to look for DE genes between conditions, there are two things we need to do:</span>
<span id="cb19-78"><a href="#cb19-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-79"><a href="#cb19-79" aria-hidden="true" tabindex="-1"></a><span class="ss">1. </span>Subset the Seurat object to our celltype of interest</span>
<span id="cb19-80"><a href="#cb19-80" aria-hidden="true" tabindex="-1"></a><span class="ss">2. </span>Set our active Idents to be the metadata column which specifies what condition each cell is</span>
<span id="cb19-81"><a href="#cb19-81" aria-hidden="true" tabindex="-1"></a>   </span>
<span id="cb19-82"><a href="#cb19-82" aria-hidden="true" tabindex="-1"></a>In our example we are focusing on vascular smooth muscle (<span class="in">`VSM`</span>) cells. The comparison we will be making is <span class="in">`TN`</span> vs. <span class="in">`cold7`</span>.</span>
<span id="cb19-83"><a href="#cb19-83" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-86"><a href="#cb19-86" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb19-87"><a href="#cb19-87" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset the object</span></span>
<span id="cb19-88"><a href="#cb19-88" aria-hidden="true" tabindex="-1"></a>seurat_vsm <span class="ot">&lt;-</span> <span class="fu">subset</span>(seurat, <span class="at">subset =</span> (celltype <span class="sc">==</span> <span class="st">"VSM"</span>))</span>
<span id="cb19-89"><a href="#cb19-89" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-90"><a href="#cb19-90" aria-hidden="true" tabindex="-1"></a><span class="co"># Set the idents</span></span>
<span id="cb19-91"><a href="#cb19-91" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(seurat_vsm) <span class="ot">&lt;-</span> <span class="st">"condition"</span></span>
<span id="cb19-92"><a href="#cb19-92" aria-hidden="true" tabindex="-1"></a>seurat_vsm</span>
<span id="cb19-93"><a href="#cb19-93" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb19-94"><a href="#cb19-94" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-95"><a href="#cb19-95" aria-hidden="true" tabindex="-1"></a>Before we make our comparisons we will explicitly set our default assay, we want to use the **normalized data, not the integrated data**.</span>
<span id="cb19-96"><a href="#cb19-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-99"><a href="#cb19-99" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb19-100"><a href="#cb19-100" aria-hidden="true" tabindex="-1"></a><span class="co"># Set default assay</span></span>
<span id="cb19-101"><a href="#cb19-101" aria-hidden="true" tabindex="-1"></a><span class="fu">DefaultAssay</span>(seurat_vsm) <span class="ot">&lt;-</span> <span class="st">"RNA"</span></span>
<span id="cb19-102"><a href="#cb19-102" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb19-103"><a href="#cb19-103" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-104"><a href="#cb19-104" aria-hidden="true" tabindex="-1"></a>The default assay should have already been <span class="in">`RNA`</span>, but we encourage you to run this line of code above to be absolutely sure in case the active slot was changed somewhere upstream in your analysis. </span>
<span id="cb19-105"><a href="#cb19-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-106"><a href="#cb19-106" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb19-107"><a href="#cb19-107" aria-hidden="true" tabindex="-1"></a><span class="fu"># Why don't we use SCT normalized data?</span></span>
<span id="cb19-108"><a href="#cb19-108" aria-hidden="true" tabindex="-1"></a>Note that the raw and normalized counts are stored in the <span class="in">`counts`</span> and <span class="in">`data`</span> slots of <span class="in">`RNA`</span> assay, respectively. By default, the functions for finding markers will use normalized data if RNA is the DefaultAssay. The number of features in the <span class="in">`RNA`</span> assay corresponds to all genes in our dataset.</span>
<span id="cb19-109"><a href="#cb19-109" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-110"><a href="#cb19-110" aria-hidden="true" tabindex="-1"></a>Now if we consider the <span class="in">`SCT`</span> assay, functions for finding markers would use the <span class="in">`scale.data`</span> slot which is the pearson residuals that come out of regularized NB regression. Differential expression on these values can be difficult interpret. Additionally, only the variable features are represented in this assay and so we may not have data for some of our marker genes.</span>
<span id="cb19-111"><a href="#cb19-111" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb19-112"><a href="#cb19-112" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-113"><a href="#cb19-113" aria-hidden="true" tabindex="-1"></a>Now we can run <span class="in">`FindMarkers()`</span>:</span>
<span id="cb19-114"><a href="#cb19-114" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-117"><a href="#cb19-117" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb19-118"><a href="#cb19-118" aria-hidden="true" tabindex="-1"></a><span class="co"># Determine differentiating markers for TN and cold7</span></span>
<span id="cb19-119"><a href="#cb19-119" aria-hidden="true" tabindex="-1"></a>dge_vsm <span class="ot">&lt;-</span> <span class="fu">FindMarkers</span>(seurat_vsm,</span>
<span id="cb19-120"><a href="#cb19-120" aria-hidden="true" tabindex="-1"></a>                       <span class="at">ident.1=</span><span class="st">"cold7"</span>,</span>
<span id="cb19-121"><a href="#cb19-121" aria-hidden="true" tabindex="-1"></a>                       <span class="at">ident.2=</span><span class="st">"TN"</span></span>
<span id="cb19-122"><a href="#cb19-122" aria-hidden="true" tabindex="-1"></a>                       )</span>
<span id="cb19-123"><a href="#cb19-123" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb19-124"><a href="#cb19-124" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-125"><a href="#cb19-125" aria-hidden="true" tabindex="-1"></a>Now let's take a quick look at the results:</span>
<span id="cb19-126"><a href="#cb19-126" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-129"><a href="#cb19-129" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb19-130"><a href="#cb19-130" aria-hidden="true" tabindex="-1"></a><span class="co"># View results </span></span>
<span id="cb19-131"><a href="#cb19-131" aria-hidden="true" tabindex="-1"></a>dge_vsm <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span>
<span id="cb19-132"><a href="#cb19-132" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb19-133"><a href="#cb19-133" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-134"><a href="#cb19-134" aria-hidden="true" tabindex="-1"></a>**The output from the `FindMarkers()` function** is a matrix containing a ranked list of differentially expressed genes listed by gene ID and associated statistics. We describe some of these columns below:</span>
<span id="cb19-135"><a href="#cb19-135" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-136"><a href="#cb19-136" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**gene:** The gene symbol</span>
<span id="cb19-137"><a href="#cb19-137" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**p_val:** The p-value not adjusted for multiple test corrections</span>
<span id="cb19-138"><a href="#cb19-138" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**avg_logFC:** The average log-fold change between sample groups. Positive values indicate that the gene is more highly expressed <span class="in">`ident.1`</span>.  </span>
<span id="cb19-139"><a href="#cb19-139" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**pct.1:** The percentage of cells where the gene is detected in <span class="in">`ident.1`</span> (cold7)    </span>
<span id="cb19-140"><a href="#cb19-140" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**pct.2:** The percentage of cells where the gene is detected on average in <span class="in">`ident.2`</span> (TN)</span>
<span id="cb19-141"><a href="#cb19-141" aria-hidden="true" tabindex="-1"></a><span class="ss">- </span>**p_val_adj:** The adjusted p-value, based on Bonferroni correction using all genes in the dataset, used to determine significance</span>
<span id="cb19-142"><a href="#cb19-142" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-143"><a href="#cb19-143" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb19-144"><a href="#cb19-144" aria-hidden="true" tabindex="-1"></a>These tests treat each cell as an independent replicate and ignore inherent correlations between cells originating from the same sample. This results in highly **inflated p-values** for each gene. Studies have been shown to find a large number of false positive associations with these results.</span>
<span id="cb19-145"><a href="#cb19-145" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb19-146"><a href="#cb19-146" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-147"><a href="#cb19-147" aria-hidden="true" tabindex="-1"></a>When looking at the output, **we suggest looking for markers with large differences in expression between `pct.1` and `pct.2` and larger fold changes**. For instance if <span class="in">`pct.1`</span> = 0.90 and <span class="in">`pct.2`</span> = 0.80, it may not be as exciting of a marker. However, if <span class="in">`pct.2`</span> = 0.1 instead, the bigger difference would be more convincing. Also of interest is whether the majority of cells expressing the gene are in the group of interest. If <span class="in">`pct.1`</span> is low, such as 0.3, it may not be as interesting. Both of these are also possible parameters to include when running the function, as described above.</span>
<span id="cb19-148"><a href="#cb19-148" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-149"><a href="#cb19-149" aria-hidden="true" tabindex="-1"></a>This is a great spot to pause and save your results!</span>
<span id="cb19-150"><a href="#cb19-150" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-153"><a href="#cb19-153" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb19-154"><a href="#cb19-154" aria-hidden="true" tabindex="-1"></a><span class="fu">write.csv</span>(dge_vsm, <span class="st">"../results/findmarkers_vsm_cold7_vs_TN.csv"</span>)</span>
<span id="cb19-155"><a href="#cb19-155" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb19-156"><a href="#cb19-156" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-157"><a href="#cb19-157" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-158"><a href="#cb19-158" aria-hidden="true" tabindex="-1"></a><span class="fu">## Significant genes</span></span>
<span id="cb19-159"><a href="#cb19-159" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-160"><a href="#cb19-160" aria-hidden="true" tabindex="-1"></a>We want to subset our results to show just our significant genes so we can begin visualizing and analyzing the results. To do this, we filter out rows based upon the <span class="in">`p_val_adj`</span> column and subsetting only genes that meet our multiple testing-corrected significance threshold of 0.05.</span>
<span id="cb19-161"><a href="#cb19-161" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-164"><a href="#cb19-164" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb19-165"><a href="#cb19-165" aria-hidden="true" tabindex="-1"></a><span class="co"># Subset significant genes</span></span>
<span id="cb19-166"><a href="#cb19-166" aria-hidden="true" tabindex="-1"></a>dge_vsm_sig <span class="ot">&lt;-</span> dge_vsm <span class="sc">%&gt;%</span> <span class="fu">subset</span>(p_val_adj <span class="sc">&lt;</span> <span class="fl">0.05</span>)</span>
<span id="cb19-167"><a href="#cb19-167" aria-hidden="true" tabindex="-1"></a>dge_vsm_sig <span class="sc">%&gt;%</span> <span class="fu">head</span>()</span>
<span id="cb19-168"><a href="#cb19-168" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb19-169"><a href="#cb19-169" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-170"><a href="#cb19-170" aria-hidden="true" tabindex="-1"></a><span class="fu">### Volcano plot</span></span>
<span id="cb19-171"><a href="#cb19-171" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-172"><a href="#cb19-172" aria-hidden="true" tabindex="-1"></a>To get a first look at the genes that are retained, we can generated a volcano plot using the <span class="in">`EnhancedVolcano()`</span> function. This is a visualization that allows us to quickly see trends in the significant genes. The x-axis here represents the average log2-fold change value, showing the degree of difference between the two conditions. On the y-axis, we see our <span class="in">`p_val_adj`</span> column represented after a negative log10 transformation is applied to better see the spread of the adjusted p-values.  </span>
<span id="cb19-173"><a href="#cb19-173" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-174"><a href="#cb19-174" aria-hidden="true" tabindex="-1"></a>Volcano plots provide a great **overview of which genes are up-regulated (positive on the x-axis) or down-regulated (negative on the x-axis)**.</span>
<span id="cb19-175"><a href="#cb19-175" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-176"><a href="#cb19-176" aria-hidden="true" tabindex="-1"></a><span class="in">```{r fig.width=10, fig.height=6}</span></span>
<span id="cb19-177"><a href="#cb19-177" aria-hidden="true" tabindex="-1"></a><span class="co"># Volcano plot</span></span>
<span id="cb19-178"><a href="#cb19-178" aria-hidden="true" tabindex="-1"></a>p_fm <span class="ot">&lt;-</span> <span class="fu">EnhancedVolcano</span>(dge_vsm_sig,</span>
<span id="cb19-179"><a href="#cb19-179" aria-hidden="true" tabindex="-1"></a>        <span class="fu">row.names</span>(dge_vsm_sig),</span>
<span id="cb19-180"><a href="#cb19-180" aria-hidden="true" tabindex="-1"></a>        <span class="at">x=</span><span class="st">"avg_log2FC"</span>,</span>
<span id="cb19-181"><a href="#cb19-181" aria-hidden="true" tabindex="-1"></a>        <span class="at">y=</span><span class="st">"p_val_adj"</span>,</span>
<span id="cb19-182"><a href="#cb19-182" aria-hidden="true" tabindex="-1"></a>       <span class="at">title=</span><span class="st">"FindMarkers VSM cells"</span>,</span>
<span id="cb19-183"><a href="#cb19-183" aria-hidden="true" tabindex="-1"></a>       <span class="at">subtitle=</span><span class="st">"TN vs cold7"</span>)</span>
<span id="cb19-184"><a href="#cb19-184" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-185"><a href="#cb19-185" aria-hidden="true" tabindex="-1"></a>p_fm</span>
<span id="cb19-186"><a href="#cb19-186" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb19-187"><a href="#cb19-187" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-188"><a href="#cb19-188" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-189"><a href="#cb19-189" aria-hidden="true" tabindex="-1"></a><span class="fu">### Violin plots</span></span>
<span id="cb19-190"><a href="#cb19-190" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-191"><a href="#cb19-191" aria-hidden="true" tabindex="-1"></a>While looking at the overall trends in the data is a great starting point, we can also start looking at genes that have large differences between <span class="in">`TN`</span> and <span class="in">`cold7`</span>. To do this, we can take a look at the top 6 genes with the smallest p-values. We additionally disregard the ribsomal genes in this visualization step.</span>
<span id="cb19-192"><a href="#cb19-192" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-195"><a href="#cb19-195" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb19-196"><a href="#cb19-196" aria-hidden="true" tabindex="-1"></a><span class="co"># Get the gene names and get the first 6 values</span></span>
<span id="cb19-197"><a href="#cb19-197" aria-hidden="true" tabindex="-1"></a><span class="co"># Ignore ribosomal genes</span></span>
<span id="cb19-198"><a href="#cb19-198" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> dge_vsm_sig <span class="sc">%&gt;%</span></span>
<span id="cb19-199"><a href="#cb19-199" aria-hidden="true" tabindex="-1"></a>  <span class="fu">rownames_to_column</span>(<span class="at">var=</span><span class="st">"gene"</span>) <span class="sc">%&gt;%</span></span>
<span id="cb19-200"><a href="#cb19-200" aria-hidden="true" tabindex="-1"></a>  <span class="fu">filter</span>(<span class="sc">!</span><span class="fu">str_detect</span>(gene, <span class="st">"Rpl|Rps"</span>)) <span class="sc">%&gt;%</span> </span>
<span id="cb19-201"><a href="#cb19-201" aria-hidden="true" tabindex="-1"></a>  <span class="fu">head</span>(<span class="dv">6</span>)</span>
<span id="cb19-202"><a href="#cb19-202" aria-hidden="true" tabindex="-1"></a>genes <span class="ot">&lt;-</span> genes<span class="sc">$</span>gene</span>
<span id="cb19-203"><a href="#cb19-203" aria-hidden="true" tabindex="-1"></a>genes</span>
<span id="cb19-204"><a href="#cb19-204" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb19-205"><a href="#cb19-205" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-206"><a href="#cb19-206" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-207"><a href="#cb19-207" aria-hidden="true" tabindex="-1"></a>With these genes selected, we can now being to visualize the distribution of expression across our two conditions using the <span class="in">`VlnPlot()`</span> function.</span>
<span id="cb19-208"><a href="#cb19-208" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-211"><a href="#cb19-211" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb19-212"><a href="#cb19-212" aria-hidden="true" tabindex="-1"></a><span class="co"># Set Idents and draw Violin plots for top 6 genes</span></span>
<span id="cb19-213"><a href="#cb19-213" aria-hidden="true" tabindex="-1"></a><span class="fu">Idents</span>(seurat_vsm) <span class="ot">&lt;-</span> <span class="st">"condition"</span></span>
<span id="cb19-214"><a href="#cb19-214" aria-hidden="true" tabindex="-1"></a><span class="fu">VlnPlot</span>(seurat_vsm, genes, <span class="at">ncol=</span><span class="dv">3</span>, <span class="at">idents=</span><span class="fu">c</span>(<span class="st">"TN"</span>, <span class="st">"cold7"</span>))</span>
<span id="cb19-215"><a href="#cb19-215" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb19-216"><a href="#cb19-216" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-217"><a href="#cb19-217" aria-hidden="true" tabindex="-1"></a><span class="fu">### UMAP plots</span></span>
<span id="cb19-218"><a href="#cb19-218" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-219"><a href="#cb19-219" aria-hidden="true" tabindex="-1"></a>When comparing two different conditions, we recommend creating a UMAP that clearly shows where the cells exist for each condition. To do so, we first need to get the UMAP coordinates for every cell of interest. When creating the scatterplot, the first thing we do is put a layer of light gray points that show the entire dataset to understand where all the cells fall. Then, we take the UMAP coordinates of the condition (<span class="in">`TN`</span> or <span class="in">`cold7`</span> in our example) and plot those on top with a color to clearly indicate where those cells are located. </span>
<span id="cb19-220"><a href="#cb19-220" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-221"><a href="#cb19-221" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb19-222"><a href="#cb19-222" aria-hidden="true" tabindex="-1"></a>This sometimes works better on the non-integrated data, so you observe a true separation of cells by condition.</span>
<span id="cb19-223"><a href="#cb19-223" aria-hidden="true" tabindex="-1"></a>:::</span>
<span id="cb19-224"><a href="#cb19-224" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-227"><a href="#cb19-227" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb19-228"><a href="#cb19-228" aria-hidden="true" tabindex="-1"></a><span class="co"># Grab the umap coordinates and condition information for each cell</span></span>
<span id="cb19-229"><a href="#cb19-229" aria-hidden="true" tabindex="-1"></a>df <span class="ot">&lt;-</span> <span class="fu">FetchData</span>(seurat_vsm, <span class="fu">c</span>(<span class="st">"umap_1"</span>, <span class="st">"umap_2"</span>, <span class="st">"condition"</span>))</span>
<span id="cb19-230"><a href="#cb19-230" aria-hidden="true" tabindex="-1"></a>df_tn <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span> <span class="fu">subset</span>(condition <span class="sc">==</span> <span class="st">"TN"</span>)</span>
<span id="cb19-231"><a href="#cb19-231" aria-hidden="true" tabindex="-1"></a>df_cold7 <span class="ot">&lt;-</span> df <span class="sc">%&gt;%</span> <span class="fu">subset</span>(condition <span class="sc">==</span> <span class="st">"cold7"</span>)</span>
<span id="cb19-232"><a href="#cb19-232" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-233"><a href="#cb19-233" aria-hidden="true" tabindex="-1"></a><span class="co"># Scatterplot of TN cells</span></span>
<span id="cb19-234"><a href="#cb19-234" aria-hidden="true" tabindex="-1"></a>p_tn <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb19-235"><a href="#cb19-235" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>df, <span class="fu">aes</span>(<span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2), <span class="at">color=</span><span class="st">"lightgray"</span>, <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">size =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb19-236"><a href="#cb19-236" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>df_tn, <span class="fu">aes</span>(<span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2), <span class="at">color=</span><span class="st">"#F8766D"</span>, <span class="at">size =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb19-237"><a href="#cb19-237" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb19-238"><a href="#cb19-238" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"VSM: TN cells"</span>)</span>
<span id="cb19-239"><a href="#cb19-239" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-240"><a href="#cb19-240" aria-hidden="true" tabindex="-1"></a><span class="co"># Scatterplot of cold7 cells</span></span>
<span id="cb19-241"><a href="#cb19-241" aria-hidden="true" tabindex="-1"></a>p_cold7 <span class="ot">&lt;-</span> <span class="fu">ggplot</span>() <span class="sc">+</span></span>
<span id="cb19-242"><a href="#cb19-242" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>df, <span class="fu">aes</span>(<span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2), <span class="at">color=</span><span class="st">"lightgray"</span>, <span class="at">alpha=</span><span class="fl">0.5</span>, <span class="at">size =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb19-243"><a href="#cb19-243" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_point</span>(<span class="at">data=</span>df_cold7, <span class="fu">aes</span>(<span class="at">x=</span>umap_1, <span class="at">y=</span>umap_2), <span class="at">color=</span><span class="st">"#00B8E7"</span>, <span class="at">size =</span> <span class="fl">0.1</span>) <span class="sc">+</span></span>
<span id="cb19-244"><a href="#cb19-244" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_classic</span>() <span class="sc">+</span></span>
<span id="cb19-245"><a href="#cb19-245" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggtitle</span>(<span class="st">"VSM: cold7 cells"</span>)</span>
<span id="cb19-246"><a href="#cb19-246" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-247"><a href="#cb19-247" aria-hidden="true" tabindex="-1"></a><span class="co"># TN and cold7 UMAPs side by side</span></span>
<span id="cb19-248"><a href="#cb19-248" aria-hidden="true" tabindex="-1"></a>p_tn <span class="sc">+</span> p_cold7</span>
<span id="cb19-249"><a href="#cb19-249" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb19-250"><a href="#cb19-250" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-251"><a href="#cb19-251" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-252"><a href="#cb19-252" aria-hidden="true" tabindex="-1"></a>This allows us to better understand our results when we look at any follow-up information on our UMAP. For example, we can begin to look at the distribution of gene expression for each of the top 6 genes with a better understanding of where the cells for each condition lie:</span>
<span id="cb19-253"><a href="#cb19-253" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-256"><a href="#cb19-256" aria-hidden="true" tabindex="-1"></a><span class="in">```{r}</span></span>
<span id="cb19-257"><a href="#cb19-257" aria-hidden="true" tabindex="-1"></a><span class="fu">FeaturePlot</span>(seurat_vsm, genes, <span class="at">ncol=</span><span class="dv">3</span>)</span>
<span id="cb19-258"><a href="#cb19-258" aria-hidden="true" tabindex="-1"></a><span class="in">```</span></span>
<span id="cb19-259"><a href="#cb19-259" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-260"><a href="#cb19-260" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-261"><a href="#cb19-261" aria-hidden="true" tabindex="-1"></a><span class="fu">## Other statistical tests</span></span>
<span id="cb19-262"><a href="#cb19-262" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-263"><a href="#cb19-263" aria-hidden="true" tabindex="-1"></a>When we looked at the extra explanations for the <span class="in">`FindMarkers()`</span> function, there was a parameter called <span class="in">`test.use`</span>. By default, the method for calculating differentially expressed genes will be a Wilcoxon Rank Sum Test. This is a fairly simple statistical approach, and there a multitude of different algorithms that can be specified. These other options are documented on the <span class="in">`FindMarkers()`</span> <span class="co">[</span><span class="ot">documentation page</span><span class="co">](https://www.rdocumentation.org/packages/Seurat/versions/5.0.3/topics/FindMarkers)</span>. For this workshop we want to highlight a few of these methods:</span>
<span id="cb19-264"><a href="#cb19-264" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-265"><a href="#cb19-265" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-266"><a href="#cb19-266" aria-hidden="true" tabindex="-1"></a><span class="fu">### Wilcoxon Rank Sum Test</span></span>
<span id="cb19-267"><a href="#cb19-267" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-268"><a href="#cb19-268" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Often described as the non-parametric version of the two-sample t-test. </span>
<span id="cb19-269"><a href="#cb19-269" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Beneficial because it can reduce the impact of outliers, which can skew the results of parametric testing.</span>
<span id="cb19-270"><a href="#cb19-270" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>It ranks the data and compares the sum of ranks within each group, to identify significant differences.</span>
<span id="cb19-271"><a href="#cb19-271" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb19-272"><a href="#cb19-272" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-273"><a href="#cb19-273" aria-hidden="true" tabindex="-1"></a><span class="fu">### DESeq2</span></span>
<span id="cb19-274"><a href="#cb19-274" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-275"><a href="#cb19-275" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Identifies differentially expressed genes between two groups of cells based on a model using DESeq2 which uses a negative binomial distribution (<span class="co">[</span><span class="ot">Love et al., 2014</span><span class="co">](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-014-0550-8)</span>). More information on DESeq2 will be provided in an upcoming lesson in this workshop.</span>
<span id="cb19-276"><a href="#cb19-276" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>This test option does not support pre-filtering of genes based on average difference (or percent detection rate) between cell groups. However, genes may be pre-filtered based on their minimum detection rate (min.pct) across both cell groups.</span>
<span id="cb19-277"><a href="#cb19-277" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-278"><a href="#cb19-278" aria-hidden="true" tabindex="-1"></a>::: callout-note</span>
<span id="cb19-279"><a href="#cb19-279" aria-hidden="true" tabindex="-1"></a>The creators of the Seurat package <span class="co">[</span><span class="ot">no longer recommend</span><span class="co">](https://github.com/satijalab/seurat/issues/2938)</span> using the <span class="in">`FindMarkers()`</span> implementation of DESeq2.</span>
<span id="cb19-280"><a href="#cb19-280" aria-hidden="true" tabindex="-1"></a>::: </span>
<span id="cb19-281"><a href="#cb19-281" aria-hidden="true" tabindex="-1"></a> </span>
<span id="cb19-282"><a href="#cb19-282" aria-hidden="true" tabindex="-1"></a><span class="fu">### MAST</span></span>
<span id="cb19-283"><a href="#cb19-283" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-284"><a href="#cb19-284" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Implements an approach that accounts for the stochastic dropout and characteristic bimodal expression distributions in which expression is either strongly non-zero or non-detectable.</span>
<span id="cb19-285"><a href="#cb19-285" aria-hidden="true" tabindex="-1"></a><span class="ss">    *  </span>A two-part, generalized linear model for such bimodal data that parameterizes both of these features    </span>
<span id="cb19-286"><a href="#cb19-286" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Also allows for estimation and control of the “cellular detection rate” (CDR) while simultaneously estimating treatment effects. This addresses the fact that cells scale transcript copy number with cell volume. </span>
<span id="cb19-287"><a href="#cb19-287" aria-hidden="true" tabindex="-1"></a><span class="ss">* </span>Permits the analysis of complex experiments, such as repeated single-cell measurements under various treatments or longitudinal sampling of single cells from multiple subjects with a variety of background characteristics (e.g., sex, age), because it can easily be extended to accommodate random effects.</span>
<span id="cb19-288"><a href="#cb19-288" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-289"><a href="#cb19-289" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb19-290"><a href="#cb19-290" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; **NOTE:** Instead of using the </span><span class="in">`FindMarkers()`</span><span class="at"> implementation, we recommend directly using the </span><span class="co">[</span><span class="ot">MAST</span><span class="co">](https://genomebiology.biomedcentral.com/articles/10.1186/s13059-015-0844-5)</span><span class="at"> algorithm from the package itself for the best results.</span></span>
<span id="cb19-291"><a href="#cb19-291" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;</span></span>
<span id="cb19-292"><a href="#cb19-292" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; If you are interested in exploring code to run MAST on this dataset directly using the package, please see the script at the link below. We recommend including the sample in the model to improve results by taking into account biological variability. Please note that this is a **computationally intensive** calculation and may take **a long time to run**.</span></span>
<span id="cb19-293"><a href="#cb19-293" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="kw">&lt;details&gt;</span></span>
<span id="cb19-294"><a href="#cb19-294" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;   </span><span class="kw">&lt;summary&gt;&lt;b&gt;&lt;i&gt;</span><span class="at">Click here for code to run MAST on this dataset side-by-side</span><span class="kw">&lt;/i&gt;&lt;/b&gt;&lt;/summary&gt;</span></span>
<span id="cb19-295"><a href="#cb19-295" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;   </span><span class="kw">&lt;br&gt;</span><span class="at">Note that this R code below uses the </span><span class="kw">&lt;b&gt;</span><span class="at">MAST library</span><span class="kw">&lt;/b&gt;</span><span class="at">. In order to run this you will need to first install the required packagea and then &gt; load the libraries.</span></span>
<span id="cb19-296"><a href="#cb19-296" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="kw">&lt;pre&gt;</span></span>
<span id="cb19-297"><a href="#cb19-297" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; library(Seurat)</span></span>
<span id="cb19-298"><a href="#cb19-298" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; library(dplyr)</span></span>
<span id="cb19-299"><a href="#cb19-299" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; library(SingleCellExperiment)</span></span>
<span id="cb19-300"><a href="#cb19-300" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; library(MAST)</span></span>
<span id="cb19-301"><a href="#cb19-301" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;</span></span>
<span id="cb19-302"><a href="#cb19-302" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="dv">&amp;#35;</span><span class="at"> Seurat to SingleCellExperiment</span></span>
<span id="cb19-303"><a href="#cb19-303" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; DefaultAssay(seurat_vsm) </span><span class="dv">&amp;lt;</span><span class="at">- "RNA"</span></span>
<span id="cb19-304"><a href="#cb19-304" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; sce &lt;- as.SingleCellExperiment(seurat_vsm)</span></span>
<span id="cb19-305"><a href="#cb19-305" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="dv">&amp;#35;</span><span class="at"> Apply log transformation</span></span>
<span id="cb19-306"><a href="#cb19-306" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; assay(sce, "logcounts") </span><span class="dv">&amp;lt;</span><span class="at">- log2(counts(sce) + 1)</span></span>
<span id="cb19-307"><a href="#cb19-307" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;</span></span>
<span id="cb19-308"><a href="#cb19-308" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="dv">&amp;#35;</span><span class="at"> Create new sce object (with only 'logcounts' matrix)</span></span>
<span id="cb19-309"><a href="#cb19-309" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; sce_1 </span><span class="dv">&amp;lt;</span><span class="at">- SingleCellExperiment(assays = list(logcounts = assay(sce, "logcounts")))</span></span>
<span id="cb19-310"><a href="#cb19-310" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; colData(sce_1) </span><span class="dv">&amp;lt;</span><span class="at">- colData(sce)</span></span>
<span id="cb19-311"><a href="#cb19-311" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span></span>
<span id="cb19-312"><a href="#cb19-312" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="dv">&amp;#35;</span><span class="at"> Change to SingleCellAssay</span></span>
<span id="cb19-313"><a href="#cb19-313" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; sca &lt;- SceToSingleCellAssay(sce_1)</span></span>
<span id="cb19-314"><a href="#cb19-314" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span></span>
<span id="cb19-315"><a href="#cb19-315" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="dv">&amp;#35;</span><span class="at"> Calculate number of genes expressed per cell and scale the value</span></span>
<span id="cb19-316"><a href="#cb19-316" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; cdr2 </span><span class="dv">&amp;lt;</span><span class="at">- colSums(SummarizedExperiment::assay(sca) </span><span class="dv">&amp;gt;</span><span class="at"> 0)</span></span>
<span id="cb19-317"><a href="#cb19-317" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; colData(sca)$cngeneson </span><span class="dv">&amp;lt;</span><span class="at">- scale(cdr2)</span></span>
<span id="cb19-318"><a href="#cb19-318" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span></span>
<span id="cb19-319"><a href="#cb19-319" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="dv">&amp;#35;</span><span class="at"> Takes a long time to calculate!</span></span>
<span id="cb19-320"><a href="#cb19-320" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="dv">&amp;#35;</span><span class="at"> Here our model includes:</span></span>
<span id="cb19-321"><a href="#cb19-321" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="dv">&amp;#9;&amp;#35;</span><span class="at"> The number of genes expressed (ngeneson)</span></span>
<span id="cb19-322"><a href="#cb19-322" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="dv">&amp;#9;&amp;#35;</span><span class="at"> Experimental condition (condition)</span></span>
<span id="cb19-323"><a href="#cb19-323" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="dv">&amp;#9;&amp;#35;</span><span class="at"> Sample as a random variable ((1 | sample))</span></span>
<span id="cb19-324"><a href="#cb19-324" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;</span></span>
<span id="cb19-325"><a href="#cb19-325" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; zlmCond &lt;- zlm(~condition + cngeneson + (1 | sample), </span></span>
<span id="cb19-326"><a href="#cb19-326" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;                sca, method="glmer", ebayes=FALSE)</span></span>
<span id="cb19-327"><a href="#cb19-327" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span></span>
<span id="cb19-328"><a href="#cb19-328" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="dv">&amp;#35;</span><span class="at"> Only test the condition coefficient.</span></span>
<span id="cb19-329"><a href="#cb19-329" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; summaryCond </span><span class="dv">&amp;lt;</span><span class="at">- summary(zlmCond, doLRT='conditionTN') </span></span>
<span id="cb19-330"><a href="#cb19-330" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span></span>
<span id="cb19-331"><a href="#cb19-331" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="dv">&amp;#35;</span><span class="at"> Some data wranging of the results</span></span>
<span id="cb19-332"><a href="#cb19-332" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; summaryDt </span><span class="dv">&amp;lt;</span><span class="at">- summaryCond$datatable</span></span>
<span id="cb19-333"><a href="#cb19-333" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; fcHurdle </span><span class="dv">&amp;lt;</span><span class="at">- merge(summaryDt</span><span class="co">[</span><span class="ot">contrast=='conditionTN' &amp; component=='H',.(primerid, `Pr(&gt;Chisq)`)</span><span class="co">]</span><span class="at">, #hurdle P values</span></span>
<span id="cb19-334"><a href="#cb19-334" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;                  summaryDt[contrast=='conditionTN' &amp; component=='logFC', </span></span>
<span id="cb19-335"><a href="#cb19-335" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;                            .(primerid, coef, ci.hi, ci.lo)], by='primerid') #logFC coefficients</span></span>
<span id="cb19-336"><a href="#cb19-336" aria-hidden="true" tabindex="-1"></a><span class="at">&gt;</span></span>
<span id="cb19-337"><a href="#cb19-337" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; fcHurdle</span><span class="co">[</span><span class="ot">,fdr:=p.adjust(`Pr(&gt;Chisq)`, 'fdr')</span><span class="co">]</span></span>
<span id="cb19-338"><a href="#cb19-338" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="kw">&lt;/pre&gt;</span></span>
<span id="cb19-339"><a href="#cb19-339" aria-hidden="true" tabindex="-1"></a><span class="at">&gt; </span><span class="kw">&lt;/details&gt;</span></span>
</code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</div></div></div></div></div>
</div> <!-- /content -->
<footer class="footer">
  <div class="nav-footer">
    <div class="nav-footer-left">
      &nbsp;
    </div>   
    <div class="nav-footer-center">This lesson has been developed by members of the teaching team at the <a href="http://bioinformatics.sph.harvard.edu/">Harvard Chan Bioinformatics Core (HBC)</a>. <br> These are open access materials distributed under the terms of the <a href="https://creativecommons.org/licenses/by/4.0/">Creative Commons Attribution license</a> (CC BY 4.0), which permits unrestricted use, distribution, and reproduction in any medium, provided the original author and source are credited. <br></div>
    <div class="nav-footer-right">
      &nbsp;
    </div>
  </div>
</footer>



</body></html>